### PR TITLE
Add OpInfo for `nn.functional.cosine_similarity`

### PR DESCRIPTION
**Pull request by krshrimali**
Please see https://github.com/facebookresearch/functorch/issues/78 and https://github.com/pytorch/pytorch/issues/54261.

Notes:

* Some redundant tests from `test_nn.py` have been removed. I'm unsure about precision checks if they can be removed as well.
* Broadcasting is also checked in the OpInfo for `cosine_similarity`.

cc: @mruberry @zou3519 @Chillee

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -1065,6 +1065,26 @@ def sample_inputs_linalg_norm(op_info, device, dtype, requires_grad):
...
+        ((), {'dim': 0}),
+        ((S, S, M), {'dim': 2}),
+        ((S, S), {})
+    )
+
+    def generator():
+        for input_shape, kwargs in cases:
+            yield SampleInput(make_arg(input_shape), args=(make_arg(input_shape),), kwargs=kwargs)
+        # Test for Broadcasting
+        yield SampleInput(make_arg((1, 2, 3)), args=(make_arg((2, 1, 3)),), kwargs={'dim': -1})
```
[the remark this answers is not in the document]
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:1085]: Thanks for the question, @zou3519 - Broadcasting was enabled for the non-reduction dimension here: https://github.com/pytorch/pytorch/pull/62912. I believe the documentation has been updated for the master as well: https://docs-preview.pytorch.org/62912/generated/torch.nn.functional.cosine_similarity.html?highlight=cosine_similarity#torch.nn.functional.cosine_similarity/.

There was a similar test added in `test_nn.py` which I moved to OpInfo. Does this answer your question?

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -1065,6 +1065,26 @@ def sample_inputs_linalg_norm(op_info, device, dtype, requires_grad):
...
         return inputs
 
+def sample_inputs_cosine_similarity(op_info, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    # Ordered as input_shape, dict of dim and eps
+    cases: Tuple[tuple, dict] = (  # type: ignore[assignment]
+        ((S, S), {'dim': 1}),
+        ((S, 2), {'dim': -1}),
+        ((S,), {'dim': 0, 'eps': 1e-5}),
```
[the remark this answers is not in the document]
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: This makes sense, thanks for pointing this out @zou3519. Has been fixed in the recent commit.

**Review by zou3519 — APPROVED** [review]

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/testing/_internal/common_methods_invocations.py:1085]: Yes, that does, thank you!

**Comment by zou3519** [comment]
I'm still working on merging this!

**Pull request closed without merging**